File: app/src/main/java/com/scichart/examples/fragments/examples2d/chartAnnotations/CustomModifierFragment.java
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2021. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// CustomModifierFragment.java is part of SCICHART®, High Performance Scientific Charts
// For full terms and conditions of the license, see http://www.scichart.com/scichart-eula/
//
// This source code is protected by international copyright law. Unauthorized
// reproduction, reverse-engineering, or distribution of all or any portion of
// this source code is strictly prohibited.
//
// This source code contains confidential and proprietary trade secrets of
// SciChart Ltd., and should at no time be copied, transferred, sold,
// distributed or made available without express written permission.
//******************************************************************************

package com.scichart.examples.fragments.examples2d.chartAnnotations;

import android.graphics.PointF;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Spinner;

import androidx.annotation.NonNull;

import com.scichart.charting.model.AnnotationCollection;
import com.scichart.charting.model.dataSeries.OhlcDataSeries;
import com.scichart.charting.modifiers.AnnotationCreationModifier;
import com.scichart.charting.modifiers.DefaultAnnotationFactory;
import com.scichart.charting.modifiers.GestureModifierBase;
import com.scichart.charting.modifiers.OnAnnotationCreatedListener;
import com.scichart.charting.visuals.ISciChartSurface;
import com.scichart.charting.visuals.SciChartSurface;
import com.scichart.charting.visuals.annotations.AnnotationCoordinateMode;
import com.scichart.charting.visuals.annotations.HorizontalAnchorPoint;
import com.scichart.charting.visuals.annotations.IAnnotation;
import com.scichart.charting.visuals.annotations.TextAnnotation;
import com.scichart.charting.visuals.annotations.TradeMarkerAnnotation;
import com.scichart.charting.visuals.annotations.VerticalAnchorPoint;
import com.scichart.charting.visuals.axes.IAxis;
import com.scichart.charting.visuals.renderableSeries.IRenderableSeries;
import com.scichart.core.framework.IUpdateSuspender;
import com.scichart.core.utility.ViewUtil;
import com.scichart.examples.components.SpinnerStringAdapter;
import com.scichart.examples.data.MarketDataService;
import com.scichart.examples.data.PriceSeries;
import com.scichart.examples.databinding.ExampleCreateAnnotationsDynamicallyFragmentBinding;
import com.scichart.examples.fragments.base.ExampleBaseFragment;
import com.scichart.examples.utils.ItemSelectedListenerBase;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Collections;
import java.util.Date;
import java.util.List;

public class CustomModifierFragment extends ExampleBaseFragment<ExampleCreateAnnotationsDynamicallyFragmentBinding> implements OnAnnotationCreatedListener {

    private final AnnotationCreationModifier annotationCreationModifier = new AnnotationCreationModifier();
    private TradeMarkerGestureModifier tradeMarkerModifier;
    private TextAnnotation instructionAnnotation;
    private final List<TextAnnotation> divAnnotations = new ArrayList<>();
    private int selectedAnnotationType = 0; // 0=Box, 1=Line, 2=TradeMarker

    @NonNull
    @Override
    protected ExampleCreateAnnotationsDynamicallyFragmentBinding inflateBinding(@NonNull LayoutInflater inflater) {
        return ExampleCreateAnnotationsDynamicallyFragmentBinding.inflate(inflater);
    }

    @Override
    protected void initExample(ExampleCreateAnnotationsDynamicallyFragmentBinding binding) {
        final Spinner annotationTypeSelector = binding.annotationTypeSelector;
        
        // Create custom array with Box, Line, and Trade Marker
        List<String> annotationTypes = Arrays.asList("BoxAnnotation", "LineAnnotation", "TradeMarker");
        annotationTypeSelector.setAdapter(new SpinnerStringAdapter(getActivity(), annotationTypes));
        annotationTypeSelector.setSelection(0);
        
        annotationTypeSelector.setOnItemSelectedListener(new ItemSelectedListenerBase() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                selectedAnnotationType = position;
                updateModifiers(binding.surface);
            }
        });

        final OhlcDataSeries<Date, Double> dataSeries = sciChartBuilder.newOhlcDataSeries(Date.class, Double.class).build();

        final MarketDataService marketDataService = new MarketDataService(Calendar.getInstance().getTime(), 5, 5);
        final PriceSeries data = marketDataService.getHistoricalData(200);

        dataSeries.append(data.getDateData(), data.getOpenData(), data.getHighData(), data.getLowData(), data.getCloseData());

        final SciChartSurface surface = binding.surface;
        
        // Use a subset of data for better zoom - show last 50 points
        final int totalPoints = dataSeries.getCount();
        final int startIndex = Math.max(0, totalPoints - 50);
        final int visibleStart = startIndex;
        final int visibleEnd = totalPoints;
        
        // Create divergence indicator data for mountain series
        final List<Integer> divAnnotationXPositions = new ArrayList<>();
        
        // Generate smooth oscillator-like data for mountain series
        for (int i = 0; i < totalPoints; i++) {
            // Mark some positions for "Div" annotations
            if (i >= startIndex && i % 25 == 0) {
                divAnnotationXPositions.add(i);
            }
        }
        
        // Add candlestick series (uses default Y-axis)
        Collections.addAll(surface.getRenderableSeries(), sciChartBuilder.newCandlestickSeries()
                .withDataSeries(dataSeries)
                .withOpacity(0.4f)
                .build());
        
        // Set visible range on X-axis to show only last 50 points
        Collections.addAll(surface.getXAxes(), sciChartBuilder.newCategoryDateAxis()
                .withVisibleRange(visibleStart, visibleEnd)
                .withGrowBy(0, 0.1)
                .build());

        // Calculate Y-axis range from visible data for better zoom
        double minPrice = Double.MAX_VALUE;
        double maxPrice = Double.MIN_VALUE;
        for (int i = startIndex; i < totalPoints; i++) {
            final double low = dataSeries.getLowValues().get(i);
            final double high = dataSeries.getHighValues().get(i);
            minPrice = Math.min(minPrice, low);
            maxPrice = Math.max(maxPrice, high);
        }
        // Add some padding
        final double priceRange = maxPrice - minPrice;
        final double yMin = minPrice - priceRange * 0.1;
        final double yMax = maxPrice + priceRange * 0.1;
        
        // Primary Y-axis for candlestick chart (default axis)
        Collections.addAll(surface.getYAxes(), sciChartBuilder.newNumericAxis()
                .withVisibleRange(yMin, yMax)
                .withGrowBy(0d, 0.1d)
                .build());
        
        // Add "Div" text annotations at specific X positions
        // Using RelativeY coordinate mode: X follows data, Y is relative (0-1)
        for (final int xPos : divAnnotationXPositions) {
            final TextAnnotation divAnnotation = sciChartBuilder.newTextAnnotation()
                    .withX1(xPos)
                    .withY1(0.95) // Position near bottom (relative Y coordinate)
                    .withCoordinateMode(AnnotationCoordinateMode.RelativeY)
                    .withText("Div")
                    .withFontStyle(12, 0xFFFFFFFF)
                    .withVerticalAnchorPoint(VerticalAnchorPoint.Bottom)
                    .withHorizontalAnchorPoint(HorizontalAnchorPoint.Center)
                    .build();
            divAnnotations.add(divAnnotation);
            surface.getAnnotations().add(divAnnotation);
        }

        // Setup annotation factory for Box and Line
        DefaultAnnotationFactory annotationFactory = new DefaultAnnotationFactory();
        annotationFactory.setFactoryForAnnotationType(DefaultAnnotationFactory.BOX_ANNOTATION, (parentSurface, annotationType) -> 
            sciChartBuilder.newBoxAnnotation().withBackgroundColor(0x6600cc00).build());
        annotationFactory.setFactoryForAnnotationType(DefaultAnnotationFactory.LINE_ANNOTATION, (parentSurface, annotationType) -> 
            sciChartBuilder.newLineAnnotation().build());

        annotationCreationModifier.setAnnotationFactory(annotationFactory);
        annotationCreationModifier.setAnnotationCreationListener(this);

        // Create trade marker modifier
        tradeMarkerModifier = new TradeMarkerGestureModifier();
        tradeMarkerModifier.setAnnotationCreationListener(this);

        updateModifiers(surface);

        binding.deleteAnnotation.setOnClickListener(v -> {
            final AnnotationCollection annotations = surface.getAnnotations();
            for (int i = annotations.size() - 1; i >= 0; i--) {
                final IAnnotation annotation = annotations.get(i);
                if (annotation.isSelected()) {
                    annotations.remove(i);
                }
            }
        });
    }

    private void updateModifiers(SciChartSurface surface) {
        // Remove all modifiers first
        surface.getChartModifiers().clear();
        
        // Remove instruction annotation if exists
        if (instructionAnnotation != null) {
            surface.getAnnotations().remove(instructionAnnotation);
            instructionAnnotation = null;
        }

        if (selectedAnnotationType == 2) {
            // Trade Marker selected - use gesture modifier and show instructions
            surface.getChartModifiers().add(sciChartBuilder.newModifierGroup().withModifier(tradeMarkerModifier).build());
            showTradeMarkerInstructions(surface);
        } else {
            // Box or Line selected - use annotation creation modifier
            annotationCreationModifier.setAnnotationType(selectedAnnotationType == 0 ? DefaultAnnotationFactory.BOX_ANNOTATION : DefaultAnnotationFactory.LINE_ANNOTATION);
            surface.getChartModifiers().add(sciChartBuilder.newModifierGroup().withModifier(annotationCreationModifier).build());
        }
    }

    private void showTradeMarkerInstructions(SciChartSurface surface) {
        instructionAnnotation = sciChartBuilder.newTextAnnotation()
                .withX1(0.01)
                .withY1(0.05)
                .withCoordinateMode(AnnotationCoordinateMode.Relative)
                .withText("- Tap to place a Buy marker\n- Double-tap to place a Sell marker")
                .withFontStyle(14, 0xFFFFFFFF)
                .withVerticalAnchorPoint(VerticalAnchorPoint.Top)
                .withHorizontalAnchorPoint(HorizontalAnchorPoint.Left)
                .build();
        
        final IUpdateSuspender suspender = surface.suspendUpdates();
        try {
            surface.getAnnotations().add(instructionAnnotation);
        } finally {
            suspender.dispose();
        }
    }

    @Override
    public void onAnnotationCreated(@NonNull IAnnotation newAnnotation) {
        newAnnotation.setIsEditable(true);
    }

    /**
     * Gesture modifier for creating trade markers with tap (buy) and double-tap (sell)
     */
    private class TradeMarkerGestureModifier extends GestureModifierBase {
        private final PointF touchPoint = new PointF();
        private OnAnnotationCreatedListener listener;

        public void setAnnotationCreationListener(OnAnnotationCreatedListener listener) {
            this.listener = listener;
        }

        @Override
        public boolean onSingleTapConfirmed(MotionEvent e) {
            createTradeMarker(e, true);
            return true;
        }

        @Override
        public boolean onDoubleTap(MotionEvent e) {
            createTradeMarker(e, false);
            return true;
        }

        private void createTradeMarker(MotionEvent e, boolean isBuy) {
            final ISciChartSurface parentSurface = getParentSurface();
            if (!(parentSurface instanceof SciChartSurface)) return;
            
            final SciChartSurface surface = (SciChartSurface) parentSurface;

            // Convert touch coordinates to chart coordinates
            touchPoint.set(e.getX(), e.getY());
            ViewUtil.translatePoint(surface.getView(), touchPoint, getModifierSurface());

            // Get axes (use default axes)
            final IAxis xAxis = surface.getXAxes().get(0);
            final IAxis yAxis = surface.getYAxes().get(0);

            if (xAxis == null || yAxis == null) return;

            // Convert to data coordinates
            final float x;
            if (xAxis.isHorizontalAxis()) {
                x = touchPoint.x;
            } else {
                x = touchPoint.y;
            }

            // Get the raw X value from touch
            final Comparable<?> rawXValue = xAxis.getDataValue(x);

            // Round to nearest integer to snap to candlestick center
            final int xIndex = Math.round(((Number) rawXValue).floatValue());

            // Use the rounded index as the final X value
            final Comparable<?> xValue = xIndex;

            // Get the candlestick data series from the renderable series
            OhlcDataSeries<?, ?> ohlcDataSeries = null;
            for (int i = 0; i < surface.getRenderableSeries().size(); i++) {
                final IRenderableSeries rs = surface.getRenderableSeries().get(i);
                if (rs.getDataSeries() instanceof OhlcDataSeries) {
                    ohlcDataSeries = (OhlcDataSeries<?, ?>) rs.getDataSeries();
                    break;
                }
            }

            // Get the Y value from the candlestick (low for buy, high for sell)
            final Comparable<?> yValue;
            if (ohlcDataSeries != null) {
                // Clamp index to valid range
                final int clampedIndex = Math.max(0, Math.min(xIndex, ohlcDataSeries.getCount() - 1));

                // Get low value for buy markers (bottom of candlestick) or high value for sell markers (top of candlestick)
                if (isBuy) {
                    yValue = ohlcDataSeries.getLowValues().get(clampedIndex);
                } else {
                    yValue = ohlcDataSeries.getHighValues().get(clampedIndex);
                }
            } else {
                // Fallback: use Y coordinate from touch if no data series found
                final float y = xAxis.isHorizontalAxis() ? touchPoint.y : touchPoint.x;
                yValue = yAxis.getDataValue(y);
            }

            final TradeMarkerAnnotation marker = new TradeMarkerAnnotation(
                    surface.getContext(),
                    xValue,
                    yValue,
                    isBuy/*,
                    quantity,
                    price,
                    change*/
            );

            // Add to surface
            final IUpdateSuspender suspender = surface.suspendUpdates();
            try {
                surface.getAnnotations().add(marker);
            } finally {
                suspender.dispose();
            }

            // Notify listener
            if (listener != null) {
                listener.onAnnotationCreated(marker);
            }
        }
    }
}

File: app/src/main/java/com/scichart/examples/fragments/examples2d/chartAnnotations/CustomModifierFragment.kt
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2021. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// CustomModifierFragment.kt is part of SCICHART®, High Performance Scientific Charts
// For full terms and conditions of the license, see http://www.scichart.com/scichart-eula/
//
// This source code is protected by international copyright law. Unauthorized
// reproduction, reverse-engineering, or distribution of all or any portion of
// this source code is strictly prohibited.
//
// This source code contains confidential and proprietary trade secrets of
// SciChart Ltd., and should at no time be copied, transferred, sold,
// distributed or made available without express written permission.
//******************************************************************************

package com.scichart.examples.fragments.examples2d.chartAnnotations.kt

import android.graphics.Color
import android.graphics.PointF
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.widget.AdapterView
import android.widget.Spinner
import com.scichart.charting.model.AnnotationCollection
import com.scichart.charting.model.dataSeries.OhlcDataSeries
import com.scichart.charting.modifiers.AnnotationCreationModifier
import com.scichart.charting.modifiers.DefaultAnnotationFactory
import com.scichart.charting.modifiers.GestureModifierBase
import com.scichart.charting.modifiers.OnAnnotationCreatedListener
import com.scichart.charting.visuals.SciChartSurface
import com.scichart.charting.visuals.annotations.AnnotationCoordinateMode
import com.scichart.charting.visuals.annotations.HorizontalAnchorPoint
import com.scichart.charting.visuals.annotations.IAnnotation
import com.scichart.charting.visuals.annotations.TextAnnotation
import com.scichart.charting.visuals.annotations.TradeMarkerAnnotation
import com.scichart.charting.visuals.annotations.VerticalAnchorPoint
import com.scichart.charting.visuals.axes.IAxis
import com.scichart.core.framework.IUpdateSuspender
import com.scichart.core.utility.ViewUtil
import com.scichart.data.model.DoubleRange
import com.scichart.examples.components.SpinnerStringAdapter
import com.scichart.examples.data.MarketDataService
import com.scichart.examples.data.PriceSeries
import com.scichart.examples.databinding.ExampleCreateAnnotationsDynamicallyFragmentBinding
import com.scichart.examples.fragments.base.ExampleBaseFragment
import com.scichart.examples.utils.ItemSelectedListenerBase
import com.scichart.examples.utils.scichartExtensions.*
import java.util.*
import kotlin.math.max
import kotlin.math.min

class CustomModifierFragment : ExampleBaseFragment<ExampleCreateAnnotationsDynamicallyFragmentBinding>(), OnAnnotationCreatedListener {

    private val annotationCreationModifier = AnnotationCreationModifier()
    private var tradeMarkerModifier: TradeMarkerGestureModifier? = null
    private var instructionAnnotation: TextAnnotation? = null
    private val divAnnotations = mutableListOf<TextAnnotation>()
    private var selectedAnnotationType = 0 // 0=Box, 1=Line, 2=TradeMarker

    override fun inflateBinding(inflater: LayoutInflater): ExampleCreateAnnotationsDynamicallyFragmentBinding {
        return ExampleCreateAnnotationsDynamicallyFragmentBinding.inflate(inflater)
    }

    override fun initExample(binding: ExampleCreateAnnotationsDynamicallyFragmentBinding) {
        val annotationTypeSelector: Spinner = binding.annotationTypeSelector
        
        // Create custom array with Box, Line, and Trade Marker
        val annotationTypes: List<String> = listOf("BoxAnnotation", "LineAnnotation", "TradeMarker")
        annotationTypeSelector.adapter = SpinnerStringAdapter(activity, annotationTypes)
        annotationTypeSelector.setSelection(0)
        
        annotationTypeSelector.onItemSelectedListener = object : ItemSelectedListenerBase() {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                selectedAnnotationType = position
                updateModifiers(binding.surface)
            }
        }

        val ohlcDataSeries: OhlcDataSeries<Date, Double> = sciChartBuilder.newOhlcDataSeries(Date::class.java, Double::class.javaObjectType).build()

        val marketDataService = MarketDataService(Calendar.getInstance().time, 5, 5)
        val data: PriceSeries = marketDataService.getHistoricalData(200)

        ohlcDataSeries.append(data.dateData, data.openData, data.highData, data.lowData, data.closeData)

        val surface: SciChartSurface = binding.surface
        
        // Use a subset of data for better zoom - show last 50 points
        val totalPoints = ohlcDataSeries.count
        val startIndex = max(0, totalPoints - 50)
        val visibleStart = startIndex.toDouble()
        val visibleEnd = totalPoints.toDouble()
        
        // Create divergence indicator data for mountain series
        val divAnnotationXPositions = mutableListOf<Int>()
        
        // Calculate Y-axis range from visible data for better zoom
        var minPrice = Double.MAX_VALUE
        var maxPrice = Double.MIN_VALUE
        for (i in startIndex until totalPoints) {
            val low = ohlcDataSeries.lowValues[i]
            val high = ohlcDataSeries.highValues[i]
            minPrice = min(minPrice, low)
            maxPrice = max(maxPrice, high)
        }
        // Add some padding
        val priceRange = maxPrice - minPrice
        val yMin = minPrice - priceRange * 0.1
        val yMax = maxPrice + priceRange * 0.1
        
        // Mark positions for "Div" annotations
        for (i in startIndex until totalPoints) {
            if (i % 25 == 0) {
                divAnnotationXPositions.add(i)
            }
        }
        
        surface.suspendUpdates {
            xAxes {
                categoryDateAxis {
                    visibleRange = DoubleRange(visibleStart, visibleEnd)
                    growBy = DoubleRange(0.0, 0.1)
                }
            }
            yAxes {
                // Primary Y-axis for candlestick chart (default axis)
                numericAxis {
                    visibleRange = DoubleRange(yMin, yMax)
                    growBy = DoubleRange(0.0, 0.1)
                }
            }
            renderableSeries {
                // Candlestick series (uses default Y-axis)
                fastCandlestickRenderableSeries {
                    dataSeries = ohlcDataSeries
                    opacity = 0.4f
                }
            }
        }
        
        // Add "Div" text annotations at specific X positions
        // Using RelativeY coordinate mode: X follows data, Y is relative (0-1)
        for (xPos in divAnnotationXPositions) {
            val divAnnotation = sciChartBuilder.newTextAnnotation()
                    .withX1(xPos)
                    .withY1(0.95) // Position near bottom (relative Y coordinate)
                    .withCoordinateMode(AnnotationCoordinateMode.RelativeY)
                    .withText("Div")
                    .withFontStyle(12f, Color.WHITE)
                    .withVerticalAnchorPoint(VerticalAnchorPoint.Bottom)
                    .withHorizontalAnchorPoint(HorizontalAnchorPoint.Center)
                    .build()
            divAnnotations.add(divAnnotation)
            surface.annotations.add(divAnnotation)
        }

        // Setup annotation factory for Box and Line
        val annotationFactory = DefaultAnnotationFactory()
        annotationFactory.setFactoryForAnnotationType(DefaultAnnotationFactory.BOX_ANNOTATION) { parentSurface, annotationType -> 
            sciChartBuilder.newBoxAnnotation().withBackgroundColor(0x6600cc00).build()
        }
        annotationFactory.setFactoryForAnnotationType(DefaultAnnotationFactory.LINE_ANNOTATION) { parentSurface, annotationType -> 
            sciChartBuilder.newLineAnnotation().build()
        }

        annotationCreationModifier.annotationFactory = annotationFactory
        annotationCreationModifier.setAnnotationCreationListener(this)

        // Create trade marker modifier
        tradeMarkerModifier = TradeMarkerGestureModifier()
        tradeMarkerModifier?.setAnnotationCreationListener(this)

        updateModifiers(surface)

        binding.deleteAnnotation.setOnClickListener { v ->
            val annotations: AnnotationCollection = surface.annotations
            for (i in annotations.size - 1 downTo 0) {
                val annotation: IAnnotation = annotations[i]
                if (annotation.isSelected) {
                    annotations.removeAt(i)
                }
            }
        }
    }

    private fun updateModifiers(surface: SciChartSurface) {
        // Remove all modifiers first
        surface.chartModifiers.clear()
        
        // Remove instruction annotation if exists
        if (instructionAnnotation != null) {
            surface.annotations.remove(instructionAnnotation)
            instructionAnnotation = null
        }

        if (selectedAnnotationType == 2) {
            // Trade Marker selected - use gesture modifier and show instructions
            surface.chartModifiers.add(sciChartBuilder.newModifierGroup().withModifier(tradeMarkerModifier!!).build())
            showTradeMarkerInstructions(surface)
        } else {
            // Box or Line selected - use annotation creation modifier
            annotationCreationModifier.annotationType = if (selectedAnnotationType == 0) DefaultAnnotationFactory.BOX_ANNOTATION else DefaultAnnotationFactory.LINE_ANNOTATION
            surface.chartModifiers.add(sciChartBuilder.newModifierGroup().withModifier(annotationCreationModifier).build())
        }
    }

    private fun showTradeMarkerInstructions(surface: SciChartSurface) {
        instructionAnnotation = sciChartBuilder.newTextAnnotation()
                .withX1(0.01)
                .withY1(0.05)
                .withCoordinateMode(AnnotationCoordinateMode.Relative)
                .withText("- Tap to place a Buy marker\n- Double-tap to place a Sell marker")
                .withFontStyle(14f, Color.WHITE)
                .withVerticalAnchorPoint(VerticalAnchorPoint.Top)
                .withHorizontalAnchorPoint(HorizontalAnchorPoint.Left)
                .build()
        
        val suspender: IUpdateSuspender = surface.suspendUpdates()
        try {
            surface.annotations.add(instructionAnnotation)
        } finally {
            suspender.dispose()
        }
    }

    override fun onAnnotationCreated(newAnnotation: IAnnotation) {
        newAnnotation.setIsEditable(true)
    }

    /**
     * Gesture modifier for creating trade markers with tap (buy) and double-tap (sell)
     */
    private inner class TradeMarkerGestureModifier : GestureModifierBase() {
        private val touchPoint = PointF()
        private var listener: OnAnnotationCreatedListener? = null

        fun setAnnotationCreationListener(listener: OnAnnotationCreatedListener?) {
            this.listener = listener
        }

        override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
            createTradeMarker(e, true)
            return true
        }

        override fun onDoubleTap(e: MotionEvent): Boolean {
            createTradeMarker(e, false)
            return true
        }

        private fun createTradeMarker(e: MotionEvent, isBuy: Boolean) {
            val parentSurface = parentSurface
            if (parentSurface == null || parentSurface !is SciChartSurface) return

            val surface = parentSurface

            // Convert touch coordinates to chart coordinates
            touchPoint.set(e.x, e.y)
            ViewUtil.translatePoint(surface.view, touchPoint, modifierSurface)

            // Get axes (use default axes)
            val xAxis: IAxis = surface.xAxes[0]
            val yAxis: IAxis = surface.yAxes[0]

            // Convert to data coordinates
            val x: Float = if (xAxis.isHorizontalAxis) {
                touchPoint.x
            } else {
                touchPoint.y
            }

            // Get the raw X value from touch
            val rawXValue = xAxis.getDataValue(x)

            // Round to nearest integer to snap to candlestick center
            val xIndex = kotlin.math.round((rawXValue as? Number)?.toFloat() ?: 0f).toInt()

            // Use the rounded index as the final X value
            val xValue: Comparable<*> = xIndex

            // Get the candlestick data series from the renderable series
            val ohlcDataSeries = surface.renderableSeries
                    .firstOrNull { it.dataSeries is OhlcDataSeries<*, *> }
                    ?.dataSeries as? OhlcDataSeries<*, *>

            // Get the Y value from the candlestick (low for buy, high for sell)
            val yValue: Comparable<*> = if (ohlcDataSeries != null) {
                // Clamp index to valid range
                val clampedIndex = max(0, min(xIndex, ohlcDataSeries.count - 1))

                // Get low value for buy markers (bottom of candlestick) or high value for sell markers (top of candlestick)
                if (isBuy) {
                    ohlcDataSeries.lowValues[clampedIndex] as Comparable<*>
                } else {
                    ohlcDataSeries.highValues[clampedIndex] as Comparable<*>
                }
            } else {
                // Fallback: use Y coordinate from touch if no data series found
                val y: Float = if (xAxis.isHorizontalAxis) touchPoint.y else touchPoint.x
                yAxis.getDataValue(y)
            }

            // Create trade marker
            val marker = TradeMarkerAnnotation(
                    surface.context,
                    xValue,
                    yValue,
                    isBuy/*,
                    quantity,
                    price,
                    change*/
            )

            // Add to surface
            val suspender: IUpdateSuspender = surface.suspendUpdates()
            try {
                surface.annotations.add(marker)
            } finally {
                suspender.dispose()
            }

            // Notify listener
            listener?.onAnnotationCreated(marker)
        }
    }
}
